ext/NormalizingFlowsEnzymeCoreExt.jl

```diff
@@ -0,0 +1,29 @@
+module NormalizingFlowsEnzymeCoreExt
+
+using EnzymeCore
+using NormalizingFlows
+using NormalizingFlows: ADTypes, DifferentiationInterface
+
+# deal with Enzyme readonly error: see https://discourse.julialang.org/t/enzyme-autodiff-readonly-error-and-working-with-batches-of-data/123012
+function NormalizingFlows._prepare_gradient(loss, adbackend::ADTypes.AutoEnzyme, θ, args...)
+    return DifferentiationInterface.prepare_gradient(
+        EnzymeCore.Const(loss),
```

Comment:
@zuhengxu, IIUC, we can do and remove deps on `EnzymeCore`. Then these functions can be moved into the main `NF` code since DI is a direct deps already. 
```suggestion
        DifferentiationInterface.Constant(loss),
```

Reply:
Not quite, the correct way to emulate this with DI alone is to change the backend object and use `AutoEnzyme(function_annotation=Enzyme.Const)`

Reply:
This doesn't quite work---maybe I'm not doing it the right way but `DI.Constant` doesn't seem to wrap the differentiated function `loss` into the right context. Here is the MWE and error message:
```julia
using NormalizingFlows
using ADTypes
using Enzyme, EnzymeCore
import DifferentiationInterface as DI

f(x, y, z) = sum(abs2, x .+ y .+ z)
T = Float64
x = randn(T, 10)
y = randn(T, 10)
z = randn(T, 10)

# wont work
at = ADTypes.AutoEnzyme(mode=Enzyme.set_runtime_activity(Enzyme.Reverse), function_annotation = Enzyme.Const)
prep = DI.prepare_gradient(DI.Constant(f), at, x, DI.Constant(y), DI.Constant(z))
value, grad = DI.value_and_gradient(DI.Constant(f), prep, at, x, DI.Constant(y), DI.Constant(z))
```

err message:
```
julia> at = ADTypes.AutoEnzyme(mode=Enzyme.set_runtime_activity(Enzyme.Reverse), function_annotation = Enzyme.Const)
AutoEnzyme(mode=ReverseMode{false, true, FFIABI, false, false}(), function_annotation=Const)

julia> prep = DI.prepare_gradient(DI.Constant(f), at, x, DI.Constant(y), DI.Constant(z))
DifferentiationInterface.NoGradientPrep()

julia> value, grad = DI.value_and_gradient(DI.Constant(f), prep, at, x, DI.Constant(y), DI.Constant(z))
ERROR: MethodError: no method matching invoke DifferentiationInterface.Constant{typeof(f)}(::Vector{Float64}, ::Vector{Float64}, ::Vector{Float64})
The type `DifferentiationInterface.Constant{typeof(f)}` exists, but no method is defined for this combination of argument types when trying to construct it.

Closest candidates are:
  (::Type{DifferentiationInterface.Constant{T}} where T)(::Any)
   @ DifferentiationInterface ~/.julia/packages/DifferentiationInterface/F5K7v/src/utils/context.jl:59

Stacktrace:
 [1] macro expansion
   @ ~/.julia/packages/Enzyme/QsaeA/src/typeutils/inference.jl:0 [inlined]
 [2] primal_return_type(mode::ReverseMode{…}, ft::Type{…}, tt::Type{…})
   @ Enzyme.Compiler ~/.julia/packages/Enzyme/QsaeA/src/typeutils/inference.jl:162
 [3] autodiff
   @ ~/.julia/packages/Enzyme/QsaeA/src/Enzyme.jl:373 [inlined]
 [4] macro expansion
   @ ~/.julia/packages/Enzyme/QsaeA/src/sugar.jl:275 [inlined]
 [5] gradient
   @ ~/.julia/packages/Enzyme/QsaeA/src/sugar.jl:263 [inlined]
 [6] value_and_gradient(::DifferentiationInterface.Constant{…}, ::DifferentiationInterface.NoGradientPrep, ::AutoEnzyme{…}, ::Vector{…}, ::DifferentiationInterface.Constant{…}, ::DifferentiationInterface.Constant{…})
   @ DifferentiationInterfaceEnzymeExt ~/.julia/packages/DifferentiationInterface/F5K7v/ext/DifferentiationInterfaceEnzymeExt/reverse_onearg.jl:217
 [7] top-level scope
   @ REPL[73]:1
Some type information was truncated. Use `show(err)` to see complete types.
```
But the following works:
```julia
# works
at2 = ADTypes.AutoEnzyme(mode=Enzyme.set_runtime_activity(Enzyme.Reverse))
prep2 = DI.prepare_gradient(EnzymeCore.Const(f), at2, x, DI.Constant(y), DI.Constant(z))
value, grad = DI.value_and_gradient(EnzymeCore.Const(f), prep, at2, x, DI.Constant(y), DI.Constant(z))
```

Reply:
Sorry for being unclear: the correct version is
```julia
at = ADTypes.AutoEnzyme(
    mode=Enzyme.set_runtime_activity(Enzyme.Reverse), 
    function_annotation = Enzyme.Const
)
prep = DI.prepare_gradient(f, at, x, DI.Constant(y), DI.Constant(z))
value, grad = DI.value_and_gradient(f, prep, at, x, DI.Constant(y), DI.Constant(z))
```
As a general rule, the only backend-specific part of a DI call should be, well, the backend object.
Furthermore, [context annotations](https://juliadiff.org/DifferentiationInterface.jl/DifferentiationInterface/stable/explanation/advanced/#Contexts) like `DI.Constant` are only used for additional arguments, not for the function itself. The `function_annotation` setting is there to ensure that `f` is wrapped into `Enzyme.Const` before being passed to `Enzyme.autodiff`, but that happens under the hood.

Reply:
ah, got it. Thank you for the clarification! I misunderstood the purpose of setting `function_annotation`. Now all tests pass nicely :)